### PR TITLE
chore: move to typescript

### PR DESCRIPTION
**Pull request by pikax**
**What kind of change does this PR introduce?** (check at least one)

- [x] Code style update
- [x] Refactor
- [x] Build-related changes
- [x] Other, please describe: Typescript code base

**Does this PR introduce a breaking change?** (check one)

- [x] No

**Other information:**

Move the code base from flow to typescript

### Plan 

This PR plans to convert the current flow code into typescript and type some of the code, avoiding changes to javascript output code.

This PR will stay open for a while, there's many pending PR which need to be merged before this. 

## Status

- [x] convert all the js files to typescript 
- [x] replace flow types with typescript
- [x] make build and CI work 
- [ ] introduce typing to plain JS files

### CODE REVIEW

File: typescript/modules.d.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      column: number
+    }): {
+      source: string | null
+      line: number | null
+      column: number | null
+    }
+  }
+}
+
+declare module 'lru-cache' {
```
Comment (KaelWD) [inline typescript/modules.d.ts:25]: Lots of these packages have `@types/<name>` available on npm

File: typescript/modules.d.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  var exports: {
+    (): any
+  }
+}
+
+declare module 'de-indent' {
+  var exports: {
+    (input: string): string
+  }
```
Comment (KaelWD) [inline typescript/modules.d.ts:34]:
```suggestion
  export default function deindent (input: string): string
```

**Comment by wyx1818** [comment]
膜拜大佬

**Comment by xiaohuahegege** [comment]
6666, Dai lo !  :smile:

**Comment by chenwuliang** [comment]
神

**Comment by gcaaa31928** [comment]
amazing PR

**Comment by ouliuquan** [comment]
改到typescript没问题，能不能跟兄弟们商量下再发PR

**Comment by wangyjx** [comment]
shocked, astonished, amazed, surprised, startled

**Comment by ifzm** [comment]
Shock ! )
This is crazy, too powerful ... 👍

**Comment by pifeifei** [comment]
人力还是 机器完成的 🔢

**Comment by NateLing** [comment]
merge,plz

**Comment by tscsds** [comment]
Merge it, quickly! It's sad it is still using the shitty "flow" what flows to nowhere.

**Comment by cereschen** [comment]
Unfortunately, there is still a lot of work to be done to achieve the perfect TS, and who is going to do that part?:sweat_smile:

**Comment by tscsds** [comment]
> Unfortunately, there is still a lot of work to be done to achieve the perfect TS, and who is going to do that part?😅

It doesn't have to be "Perfect TS", just replacing the typing with TS is already enough, much better for reading the code.

**Comment by cereschen** [comment]
> > Unfortunately, there is still a lot of work to be done to achieve the perfect TS, and who is going to do that part?😅
> 
> It doesn't have to be "Perfect TS", just replacing the typing with TS is already enough, much better for reading the code.
I agree with you. At least it's better than before

**Comment by JdesEva** [comment]
6666

**Comment by XinChou16** [comment]
awesome👏

**Comment by ytianzaishui** [comment]
What a nice job!

**Comment by dinhquochan** [comment]
Epic PR 🤯

**Comment by myluke** [comment]
路过，看看这个牛逼的pr

**Comment by wellfrog16** [comment]
历史性的PR，赶紧前排占个位置。

**Comment by GodLyons** [comment]
这样的话vue代码的可读性将会提升n个台阶

**Comment by dargoner** [comment]
膜拜

**Comment by nabaonan** [comment]
膜拜大佬，不知会不会在2.7中发布，期待merge

**Review by iceman201 — APPROVED** [review]

**Comment by tscsds** [comment]
The PR is approved! Is it going to be merged soon?

**Comment by 442623641** [comment]
vue + TypeScript development tools, really good, greatly enhance the code maintenance and development efficiency
[vue-ts-toolkit](https://github.com/442623641/vue-ts-toolkit)

**Comment by xieyezi** [comment]
It's so cool, But there no people to review this.😄

**Comment by zxenonx** [comment]
> It's so cool, But No one can review it.😄

Really? 😅

**Comment by iDestin** [comment]
cool！！！ 😄

**Comment by sondh0127** [comment]
holy god!

**Comment by yyx990803** [comment]
To avoid merge issues w/ `dev`, I created a new `main` branch which is based on this PR - thanks a lot for the heroic work!

**Pull request closed without merging**